abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1969 @@
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) num_trailers`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called when HTTP request trailers are received from downstream.
+
+All `num_trailers` trailers can be retrieved and/or modified using
+[`proxy_get_header_map_pairs`] and/or [`proxy_set_header_map_pairs`]
+with `map_id` set to `HTTP_REQUEST_TRAILERS`.
+
+Individual HTTP request trailers can be retrieved and/or modified using
+[`proxy_get_header_map_value`], [`proxy_replace_header_map_value`],
+[`proxy_add_header_map_value`] and/or [`proxy_remove_header_map_value`]
+with `map_id` set to `HTTP_REQUEST_TRAILERS`.
+
+Paused request can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_REQUEST`.
+
+Additionally, as long as HTTP response headers were not send downstream,
+a HTTP response can be send using [`proxy_send_local_response`].
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_REQUEST_TRAILERS` fields downstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_response_headers`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) num_headers`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called when HTTP response headers are received from upstream.
+
+All `num_headers` headers can be retrieved and/or modified using
+[`proxy_get_header_map_pairs`] and/or [`proxy_set_header_map_pairs`]
+with `map_id` set to `HTTP_RESPONSE_HEADERS`.
+
+Individual headers can be retrieved and/or modified using
+[`proxy_get_header_map_value`], [`proxy_replace_header_map_value`],
+[`proxy_add_header_map_value`] and/or [`proxy_remove_header_map_value`]
+with `map_id` set to `HTTP_RESPONSE_HEADERS`.
+
+Paused request can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_RESPONSE`.
+
+Additionally, instead of forwarding the response from upstream,
+a new HTTP response can be send using [`proxy_send_local_response`].
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_RESPONSE_HEADERS` fields downstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_response_body`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) body_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called for each chunk of HTTP response body received from upstream.
+
+Response body chunk (of `body_size`) can be retrieved and/or modified
+using [`proxy_get_buffer_bytes`] and/or [`proxy_set_buffer_bytes`]
+with `buffer_id` set to `HTTP_RESPONSE_BODY`.
+
+Paused HTTP response can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_RESPONSE`.
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_RESPONSE_BODY` buffer downstream.
+- `PAUSE` to pause processing.
```

Comment:
In the Rust SDK examples we recently noticed that returning `PAUSE` expects the host implementation to buffer the response body until `eof`, at which point it will call `on_response_body` once more (with a full body and `eof=true`). We saw this described [in this example](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/master/examples/http_body/src/lib.rs#L53-L55) but not written explicitly so anywhere else. Would this be the right place to clarify this behavior?

Reply:
> We could add a note that calling proxy_on_{request,response}_body with the same buffer size twice in a row implies that the buffer is full, but that's a bit hacky and not currently implemented.

I suppose that is not what the proxy-wasm-rust-sdk example expects, and I don't know how much flexibility there is for v0.2.1 and this corner-case, but what if we made it clear body handlers can always be invoked with more chunks so long as `eof: false`? In short doing something like `vNEXT` `WaitForEndOrFull`, but implied.

1. Handlers return `again` the first time which enables buffering.
2. The same handler expects another invocation with a chunk that may or may not fit in the body buffer.
3. Next body handler invocation:
  3a. If `eof: true`, the full body fits in the buffer on this invocation.
  3b. If `eof: false`, the buffer may or may not be full, but either way the handler can consider this body to only be handled by chunks, and expect more invocations.

In this context, hosts are expected to print appropriate error logs for a full body buffer or a second `again` return value. Filters are them expected to keep track of body handlers invocations themselves on a per-need basis. Or rather, body handlers are expected to always handle chunked bodies as a fail-safe.

Maybe this misses cases I haven't thought of, or maybe it matches existing implementations, I don't know. Plus of course I am unsure how much flexibility there is for existing host and filters of v0.2.1.

By the way, what would happen to the [proxy-wasm-rust-sdk example](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/master/examples/http_body/src/lib.rs#L52-L56) on Envoy, would the body not fit the buffer? I ought to check but I don't know my way around Envoy very well...

Reply:
Shall the actions need to extend a bit and make it more explicit? the Pause will actually convert into `Http::FilterDataStatus::StopIterationAndBuffer` status implicitly which will buffer the response body. more context: https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/414